[firebase-analytics] Add pure JS analytics for use on Expo Client

IjzerenHein opened the pull request:
# Why

This PR adds a Pure JS Firebase Analytics module to `expo-firebase-analytics`, which is used on the standard Expo Client to log events. Using Firebase Analytics is otherwise not possible as the firebase JS SDK Analytics module requires browser features, and the native Firebase SDK requires a google-services file to be bundled with the app, and cannot be  (re-)initialized correctly.

The pure JS implementation makes it possible for users to use and test firebase analytics without having to create a standalone build or use a custom client.

# How

A Pure JS Firebase Analytics implementation was created which uses the Google Measurement Protocol version 2. This protocol is used by the gtag.js/firebase-sdk itself and communicates with the `https://www.google-analytics.com` domain. This protocol is not officially documented by Google yet, so it may be affected by API changes in the future.
In order to log events to the Google Analytics server, a `measurementId` is needed. This can be configured in `app.json`, [through the new `web.config.firebase` key.](#7086) 

# Test Plan

Some manual steps need to be performed in order to test this using the test-suite.
I've been working on getting google-services config dynamically linked in test-suite using `app.config.js` but that is not ready yet and requires updates to expo-cli.

- Follow the steps [to setup "native" firebase with the test-suite](#6662)
- Make sure to include the web-configuration
- Open the project on the standard Expo Client
- Run the "Firebase Analytics" tests
- After a short while (couple mins at most) the Analytics events show up in your firebase project

IjzerenHein committed:
[firebase-analytics] WIP js analytics

IjzerenHein committed:
[firebase-analytics] Add Pure JS FirebaseAnalytics (wip, 80% done)

IjzerenHein committed:
[firebase-analytics] Add pure JS Firebase Analytics (done)

IjzerenHein committed:
[firebase-analytics] Update & integrate Pure JS Analytics tracker

IjzerenHein committed:
[docs] Add instruction on how to use firebase analytics on Expo Client

brentvatne commented:
> A Pure JS Firebase Analytics implementation was created which uses the Google Measurement Protocol version 2. This protocol is used by the gtag.js/firebase-sdk itself and communicates with the https://www.google-analytics.com domain. This protocol is not officially documented by Google yet, so it may be affected by API changes in the future.

It might be worth reaching out to the Firebase team and letting them know that we plan on using this and what for, just so that they are aware. That said, given that it's used by those gtag.js/firebase-sdk I doubt they will make abrupt breaking changes without reasonable warning / chance to upgrade.

brentvatne commented on docs/pages/versions/unversioned/sdk/firebase-analytics.md:
```diff
@@ -37,10 +37,28 @@ The use of Native Firebase Analytics requires that the google-services configura
-by calling `setUnavailabilityLogging(false)`.
+Instead, the standard Expo Client relies on a pure JavaScript implementation of Firebase Analytics
+to log events. This means that certain native lifecyle events are not recorded in the standard client,
+but you can still use `logEvent` to log events.
```
```suggestion
but you can still use `logEvent` to record events.
```

brentvatne commented on docs/pages/versions/unversioned/sdk/firebase-analytics.md:
```diff
@@ -37,10 +37,28 @@ The use of Native Firebase Analytics requires that the google-services configura
-purposes. It also generates a one time warning to let you know that analytics events are not recorded
-on the standard Expo Client. You can suppress this warning and disable the logging
-by calling `setUnavailabilityLogging(false)`.
+Instead, the standard Expo Client relies on a pure JavaScript implementation of Firebase Analytics
```
```suggestion
Instead, the standard Expo client relies on a JavaScript-based implementation of Firebase Analytics
```

brentvatne commented on docs/pages/versions/unversioned/sdk/firebase-analytics.md:
```diff
@@ -37,10 +37,28 @@ The use of Native Firebase Analytics requires that the google-services configura
+
+In order to use Firebase Analytics on the standard Client, ensure that the firebase web-configuration
+is set in `app.json` and that `measurementId` exists in your firebase config. If `measurementId` 
+doesn't exist, then you need to enable or update Google Analytics in your Firebase project.
```
collapse paragraphs into one line so we don't add unnecessary linebreaks to docs

brentvatne commented on docs/pages/versions/unversioned/sdk/firebase-analytics.md:
```diff
@@ -37,10 +37,28 @@ The use of Native Firebase Analytics requires that the google-services configura
+to log events. This means that certain native lifecyle events are not recorded in the standard client,
+but you can still use `logEvent` to log events.
+
+In order to use Firebase Analytics on the standard Client, ensure that the firebase web-configuration
```
```suggestion
You may want to use Firebase Analytics in Expo client to verify that you are logging events at the time you intend to and with the data that you want to attach without having to do a standalone app build. To do set this up, ensure that the Firebase web configuration
```

brentvatne commented on docs/pages/versions/unversioned/sdk/firebase-analytics.md:
````diff
@@ -37,10 +37,28 @@ The use of Native Firebase Analytics requires that the google-services configura
+**app.json**
+```json
+{
+  "web": {
````
might want to nest this under the `"expo"` key to be extra clear

brentvatne commented on packages/expo-firebase-analytics/src/ExpoFirebaseAnalytics.ts:
```diff
@@ -31,10 +35,22 @@ function callAnalyticsModule(funcName: string, ...args) {
+      pureJSAnalyticsTracker = new FirebaseAnalyticsJS(defaultWebOptions, {
+        clientId: Constants.sessionId,
+        strictNativeEmulation: true,
+        appName: Constants.manifest?.name || 'Expo Client',
```
```suggestion
        appName: Constants.manifest?.name || 'Unnamed Expo project',
```

brentvatne commented on packages/expo-firebase-analytics/src/ExpoFirebaseAnalytics.ts:
```diff
@@ -31,10 +35,22 @@ function callAnalyticsModule(funcName: string, ...args) {
       if (!isUnavailabilityWarningLogged) {
         console.warn(
-          'Firebase Analytics is not available in the Expo client. To test Firebase Analytics create a stand-alone build or custom client. To suppress this warning use `setUnavailabilityLogging(false)`.'
+          `Firebase Analytics is not available in the Expo client. See "https://docs.expo.io/versions/unversioned/sdk/firebase-analytics#expo_client_limitations_configuration" on more information on setting up Firebase Analytics with the standard Expo Client.`
```
```suggestion
          `Firebase Analytics is not available in the Expo client. See "https://docs.expo.io/versions/latest/sdk/firebase-analytics" on more information on setting up Firebase Analytics with the standard Expo Client.`
```

- link to latest rather than unversioned
- leave out anchor as those get stale pretty quick and we don't have any automated validation for links from code

brentvatne commented on packages/expo-firebase-analytics/src/FirebaseAnalyticsJS.ts:
```diff
@@ -0,0 +1,319 @@
+      };
+    }
+    const args = encodeQueryArgs(queryArgs);
+    console.log(`FirebaseAnalyticsJS: ${args}...`);
```
remove logs

IjzerenHein commented:
> > A Pure JS Firebase Analytics implementation was created which uses the Google Measurement Protocol version 2. This protocol is used by the gtag.js/firebase-sdk itself and communicates with the https://www.google-analytics.com domain. This protocol is not officially documented by Google yet, so it may be affected by API changes in the future.
> 
> It might be worth reaching out to the Firebase team and letting them know that we plan on using this and what for, just so that they are aware. That said, given that it's used by those gtag.js/firebase-sdk I doubt they will make abrupt breaking changes without reasonable warning / chance to upgrade.

I've posted an issue on the firebase-js-sdk repo:
https://github.com/firebase/firebase-js-sdk/issues/2644

IjzerenHein committed:
[firebase-analytics] Fixes after review feedback

IjzerenHein committed:
[firebase-analytics] Update JS build files

IjzerenHein committed:
[firebase-analytics] Update analytics-js to use DEFAULT_WEB_APP_OPTIONS

IjzerenHein committed:
[firebase-analytics] Update JS build files

IjzerenHein committed:
[firebase-analytics] Add recording of `screen_view` events on Expo client, when calling `setCurrentScreen`

This mimics the behavior on Native, when calling setCurrentScreen also records a screen-view event.

Note that this is different from the Firebase JS SDK, which does not record a screen-view event, and requires `logEvent` to be called manually.

IjzerenHein committed:
[firebase-analytics] Update JS build files

brentvatne commented:
it seems like there is some risk that the measurement protocol v2 will change without warning, but i think this is fine:

- this would only impact expo client and we can roll out and update to this library easily since it's just js
- we have a dialogue open with firebase to get more info about it and also for supporting multiple apps with native analytics
- it's the best workaround that we can think of so far
